### PR TITLE
[GXA-3518] feat: add Stratified splitters

### PR DESCRIPTION
**Pull request by ludovico-lanni**

### CODE REVIEW

File: tests/test_splitter.py [hunk 1]
```diff
@@ -106,3 +149,103 @@ def test_switchback_balanced_splitter_abc():
         counts = sampled_treatment.treatment.value_counts()
         assert counts.max() == 2
         assert counts.min() == 1
+
+
+def test_stratified(df_strata):
+    splitter = StratifiedClusteredSplitter(
```
Comment (david26694) [inline tests/test_splitter.py:155]: There is a lot of code repetition in here, we could extract a function cllaed assert_balanced_strata(df) and just call this function in the tests

File: cluster_experiments/random_splitter.py [hunk 2]
```diff
@@ -198,3 +193,59 @@ def assign_treatment_df(
...
+            if df_unique.groupby(self.cluster_cols)[strata_col].nunique().max() > 1:
+                raise ValueError(
+                    f"There are multiple values in {strata_col} for the same cluster item"
+                    "You cannot stratify on this column"
+                )
+
+        # random shuffling
+        random_sorted_treatments = list(np.random.permutation(self.treatments))
+
+        df_unique_shuffled = df_unique.sample(frac=1)
```
Comment (david26694) [inline cluster_experiments/random_splitter.py:228]: Couple things: you could do this in line 215, in a more pipeline manner, and also do a reset_index(drop=True) after this to have less noise with indices

Comment (ludovico-lanni) [reply cluster_experiments/random_splitter.py]: done!

File: tests/test_splitter.py [hunk 3]
```diff
@@ -106,3 +149,103 @@ def test_switchback_balanced_splitter_abc():
...
+    for treatment in ["A", "B"]:
+        for segment in ["good", "bad"]:
+            assert (
+                treatment_df.query(f"treatment == '{treatment}'")["segment"]
+                .value_counts(normalize=True)[segment]
+                .squeeze()
+            ) == 0.5
+
+
+def test_stratified_shuffled_input(df_strata):
```
[the remark this answers is not in the document]
Comment (ludovico-lanni) [reply tests/test_splitter.py]: done! added for other tests as well

File: tests/test_splitter.py (hunk 1 above)
Comment (ludovico-lanni) [reply tests/test_splitter.py:177]: done! changed the assertion pipeline to test clusters-strata unique combinations, as this is what the splitter does.

File: tests/test_splitter.py [hunk 4]
```diff
@@ -32,6 +33,48 @@ def df_cluster(clusters):
     )
 
 
+@pytest.fixture
```
[the remark this answers is not in the document]
Comment (ludovico-lanni) [reply tests/test_splitter.py:36]: done!

File: cluster_experiments/random_splitter.py [hunk 5]
```diff
@@ -198,3 +193,59 @@ def assign_treatment_df(
     def from_config(cls, config):
         """Creates a NonClusteredSplitter from a PowerConfig"""
         return cls(treatments=config.treatments, treatment_col=config.treatment_col)
+
+
+class StratifiedClusteredSplitter(RandomSplitter):
```
[the remark this answers is not in the document]
Comment (ludovico-lanni) [reply cluster_experiments/random_splitter.py:198]: done!

**Comment by ludovico-lanni** [comment]
- refactored code as suggested
- update readme and init files with new class
- changed the logic of the tests as we were testing for balance in the number of rows across strata, instead of number of clusters across strata

File: tests/test_splitter.py [hunk 6]
```diff
@@ -106,3 +168,138 @@ def test_switchback_balanced_splitter_abc():
...
+        strata_cols=strata_cols, cluster_cols=cluster_cols
+    )
+
+    treatment_df = splitter.assign_treatment_df(df)
+
+    treatment_df_unique = treatment_df[
+        strata_cols + cluster_cols + ["treatment"]
+    ].drop_duplicates()
+
+    ##
```
Comment (david26694) [inline tests/test_splitter.py:187]: remove unneeded #

File: tests/test_splitter.py [hunk 7]
```diff
@@ -106,3 +168,138 @@ def test_switchback_balanced_splitter_abc():
...
+    if len(cluster_cols) > 1:
```
Comment (david26694) [inline tests/test_splitter.py:188]: can we extract this into a function? the whole if: agg else

File: tests/test_splitter.py [hunk 8]
```diff
@@ -106,3 +168,138 @@ def test_switchback_balanced_splitter_abc():
...
+        for stratus in treatment_df_unique["strata_test"].unique():
+            assert (
+                treatment_df_unique.query(f"strata_test == '{stratus}'")["treatment"]
+                .value_counts(normalize=True)[treatment]
+                .squeeze()
+            ) == 1 / len(treatments)
+
+
+def test_stratified(df_strata):
+
```
Comment (david26694) [inline tests/test_splitter.py:213]: style thing: no need for a line in here

File: tests/test_splitter.py [hunk 9]
```diff
@@ -106,3 +168,138 @@ def test_switchback_balanced_splitter_abc():
...
+    # base test for balance of treatments across strata
+
```
Comment (david26694) [inline tests/test_splitter.py:215]: style thing: no need for a line in here

**Review by david26694 — APPROVED** [review]
could you change these two minor things and merge afterwards?

File: tests/test_splitter.py (hunk 6 above)
Comment (ludovico-lanni) [reply tests/test_splitter.py:187]: done!

File: tests/test_splitter.py (hunk 8 above)
Comment (ludovico-lanni) [reply tests/test_splitter.py:213]: done!

File: tests/test_splitter.py (hunk 9 above)
Comment (ludovico-lanni) [reply tests/test_splitter.py:215]: done!

File: tests/test_splitter.py (hunk 7 above)
Comment (ludovico-lanni) [reply tests/test_splitter.py]: done! `combine_columns(df, cols_list)`

**Comment by ludovico-lanni** [comment]
- fixed minor style inaccuracies
- extracted function to combine columns

**Pull request merged**